dotnet/src/Planners/Planners.Handlebars/Handlebars/HandlebarsPlanner.cs

```diff
@@ -79,33 +79,49 @@ public Task<HandlebarsPlan> CreatePlanAsync(Kernel kernel, string goal, KernelAr
-        }
+        string? createPlanPrompt = null;
+        ChatMessageContent? modelResults = null;
 
-        Match match = Regex.Match(completionResults.Content, @"```\s*(handlebars)?\s*(.*)\s*```", RegexOptions.Singleline);
-        if (!match.Success)
+        try
         {
-            throw new KernelException($"[{HandlebarsPlannerErrorCodes.InvalidTemplate}] Could not find the plan in the results\nPlanner output:\n{completionResults}");
-        }
+            // Get CreatePlan prompt template
+            var functionsMetadata = await kernel.Plugins.GetFunctionsAsync(this._options, null, null, cancellationToken).ConfigureAwait(false);
+            var availableFunctions = this.GetAvailableFunctionsManual(functionsMetadata, out var complexParameterTypes, out var complexParameterSchemas);
+            createPlanPrompt = await this.GetHandlebarsTemplateAsync(kernel, goal, arguments, availableFunctions, complexParameterTypes, complexParameterSchemas, cancellationToken).ConfigureAwait(false);
+            ChatHistory chatMessages = this.GetChatHistoryFromPrompt(createPlanPrompt);
+
+            // Get the chat completion results
+            var chatCompletionService = kernel.GetRequiredService<IChatCompletionService>();
+            modelResults = await chatCompletionService.GetChatMessageContentAsync(chatMessages, executionSettings: this._options.ExecutionSettings, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            // Check if plan could not be created due to insufficient functions
+            if (modelResults.Content is not null && modelResults.Content.IndexOf(InsufficientFunctionsError, StringComparison.OrdinalIgnoreCase) >= 0)
+            {
+                var functionNames = availableFunctions.ToList().Select(func => $"{func.PluginName}{this._templateFactory.NameDelimiter}{func.Name}");
+                throw new KernelException(
+                    $"[{HandlebarsPlannerErrorCodes.InsufficientFunctionsForGoal}] Unable to create plan for goal with available functions.");
+            }
 
-        var planTemplate = match.Groups[2].Value.Trim();
-        planTemplate = MinifyHandlebarsTemplate(planTemplate);
+            Match match = Regex.Match(modelResults.Content, @"```\s*(handlebars)?\s*(.*)\s*```", RegexOptions.Singleline);
+            if (!match.Success)
+            {
+                throw new KernelException($"[{HandlebarsPlannerErrorCodes.InvalidTemplate}] Could not find the plan in the results.");
+            }
+
+            var planTemplate = match.Groups[2].Value.Trim();
+            planTemplate = MinifyHandlebarsTemplate(planTemplate);
 
-        return new HandlebarsPlan(planTemplate, createPlanPrompt);
+            return new HandlebarsPlan(planTemplate, createPlanPrompt);
+        }
+        catch (KernelException ex)
+        {
+            throw new HandlebarsPlanCreationException(
```

Comment:
I am just wondering how useful the "CreatePlan failed. See inner exception for details." message and exception are from the code consumer's point of view. The consumer will have to catch it and use if/switch logic inside to test the internal exception type/details to make a recovery decision.

Wouldn't it be easier for the consumer to catch a more specific exception like UnachievableGoalException with a message, goal, list of functions, and model results?

@stephentoub what do you think?

Reply:
It's set up this way so that `createPlanPrompt` and `modelResults`, if available, would be populated into all exceptions, accounting for unhandled KernelExceptions. Otherwise, we could just override the KernelException and port the message over to this `PlanCreationException`:

```
throw new PlanCreationException(
                ex.Message,
                createPlanPrompt,
                modelResults
            );
```

The CreatePlanPrompt should contain all the necessary details like list of functions and goal and this field also captures the scenario where a user provides their own custom planner prompt to override the default